cli/config/addressbook.go

```diff
@@ -0,0 +1,169 @@
+	ethAddressCfgKey = "eth_address"
+)
+
+// AddressBook contains the configuration of the address book.
+type AddressBook struct {
+	// All is a map of all configured address entries in the address book.
+	All map[string]*AddressBookEntry `mapstructure:",remain"`
+}
+
+// Validate performs config validation.
+func (ab *AddressBook) Validate() error {
+	// Make sure all entries are valid.
+	for name, acc := range ab.All {
+		if err := config.ValidateIdentifier(name); err != nil {
+			return fmt.Errorf("malformed address name '%s': %w", name, err)
+		}
+
+		if err := acc.Validate(); err != nil {
+			return fmt.Errorf("address '%s': %w", name, err)
+		}
+	}
+
+	return nil
+}
+
+// Remove removes the given address book entry.
+func (ab *AddressBook) Remove(name string) error {
+	if _, exists := ab.All[name]; !exists {
+		return fmt.Errorf("address named '%s' does not exist in the address book", name)
+	}
+
+	if err := config.ValidateIdentifier(name); err != nil {
+		return fmt.Errorf("malformed address name '%s': %w", name, err)
+	}
+
+	delete(ab.All, name)
+
+	return nil
+}
+
+// Rename renames an existing address book entry.
+func (ab *AddressBook) Rename(old, new string) error {
+	cfg, exists := ab.All[old]
+	if !exists {
+		return fmt.Errorf("address named '%s' does not exist", old)
+	}
+
+	if _, exists = ab.All[new]; exists {
+		return fmt.Errorf("address named '%s' already exists", new)
+	}
+
+	if err := config.ValidateIdentifier(old); err != nil {
+		return fmt.Errorf("malformed old address name '%s': %w", old, err)
+	}
+	if err := config.ValidateIdentifier(new); err != nil {
+		return fmt.Errorf("malformed new address name '%s': %w", new, err)
+	}
+
+	ab.All[new] = cfg
+	delete(ab.All, old)
+
+	return nil
+}
+
+// Add adds new address book entry.
+func (ab *AddressBook) Add(name string, address string) error {
+	if _, exists := ab.All[name]; exists {
+		return fmt.Errorf("address named '%s' already exists in the address book", name)
+	}
+
+	if err := config.ValidateIdentifier(name); err != nil {
+		return fmt.Errorf("malformed address name '%s': %w", name, err)
+	}
+
+	nativeAddr, ethAddr, err := helpers.ResolveEthOrOasisAddress(address)
+	if err != nil {
+		return err
+	}
+
+	nativeAddrStr, err := nativeAddr.MarshalText()
+	if err != nil {
+		return fmt.Errorf("failed to marshal address %w", err)
+	}
+
+	abEntry := AddressBookEntry{
+		Address: string(nativeAddrStr),
+		Config:  map[string]interface{}{},
+	}
+
+	if ethAddr != nil {
+		abEntry.Config[ethAddressCfgKey] = ethAddr.Hex()
+	}
+
+	if ab.All == nil {
+		ab.All = make(map[string]*AddressBookEntry)
+	}
+	ab.All[name] = &abEntry
+
+	return nil
+}
+
+// AddressBookEntry is a configuration object for a single entry in the address book.
+type AddressBookEntry struct {
+	Description string `mapstructure:"description"`
+	Address     string `mapstructure:"address"`
+
+	// Config contains address-specific configuration for this entry.
+	Config map[string]interface{} `mapstructure:",remain"`
```

Comment:
This `Config` field can be removed now (for wallet accounts it is there because each account backend can have its own configuration) and you can just make a field called `EthAddress` directly in the struct (with `omitempty`).